### PR TITLE
run: Prevent argument duplication with deps/outs

### LINKED ISSUE #1168: run: detect output duplication in the command itself
I.e.:
```
(dvc) ➜  myrepo git:(master) dvc run -o foo -M foo 'echo foo > foo'
Running command:
	echo foo > foo
Adding 'foo' to '.gitignore'.
Saving 'foo' to cache '.dvc/cache'.
Created 'hardlink': .dvc/cache/d3/b07384d113edec49eaa6238ad5ff00 -> foo
Saving information to 'foo.dvc'.

To track the changes with git run:

	git add .gitignore foo.dvc
(dvc) ➜  myrepo git:(master) ✗ dvc status
Error: Failed to obtain data status: File '/Users/efiop/myrepo/foo' is specified as an output in more than one stage:
    foo.dvc
    foo.dvc

```

### LINKED ISSUE #1255: run: Warn about duplicated arguments
Summary
-------------
You can specify the same output more than one type without any hint.
It would be great to have at least a warning, helping the user with potential typos.

Expected behavior
------------------------
```bash
dvc run \
  --file double-output.dvc \
  --outs hello.txt \
  --outs hello.txt \
  'echo "hello" > hello.txt'

Warning: Output 'hello.txt' was specified more than one time
```

`cat double-output.dvc`:
```
cmd: echo "hello" > hello.txt
md5: ec93debb48f6f21a03f5dbe10fa0a6c2
outs:
- cache: true
  md5: b1946ac92492d2347c6235b4d2611184
  path: hello.txt
```

Actual behavior
--------------------
```bash
dvc run \
  --file double-output.dvc \
  --outs hello.txt \
  --outs hello.txt \
  'echo "hello" > hello.txt'
```

`cat double-output.dvc`:
```
cmd: echo "hello" > hello.txt
md5: ec93debb48f6f21a03f5dbe10fa0a6c2
outs:
- cache: true
  md5: b1946ac92492d2347c6235b4d2611184
  path: hello.txt
- cache: true
  md5: b1946ac92492d2347c6235b4d2611184
  path: hello.txt
```

System configuration
----------------------------
DVC version: `0.19.12`
Installation: `Arch Linux AUR package`

### PR DESCRIPTION
**Pull request by mroutis**
Fix #1255
Fix #1168
Close #1259

RFC:
- Add / fix docstrings and privatize methods
- Add __repr__ method for core objects

### CODE REVIEW

File: dvc/state.py [hunk 1]
```diff
@@ -69,20 +69,29 @@ def __init__(self, project, config):
         self.inserts = 0
 
     def __enter__(self):
-        self.load()
+        if not self.c:
```
Comment (mroutis) [inline dvc/state.py]: Sorry, @efiop , this is legacy... Ignore this PR while I fix it

**Comment by mroutis** [comment]
@efiop , what do you think about moving this checks to `Stage.loads` ?

https://github.com/mroutis/dvc/blob/8ef7a324e0fde8fea12343360ed844268c47b3d4/dvc/project.py#L195-L213

**Comment by efiop** [comment]
> @efiop , what do you think about moving this checks to Stage.loads ?
> https://github.com/mroutis/dvc/blob/8ef7a324e0fde8fea12343360ed844268c47b3d4/dvc/project.py#L195-L213

I don't think that would work. Note that those methods need a list of loaded stages, and we also use loads() while loading those stages :slightly_smiling_face: We should check for such errors both on `dvc run`(to avoid creating bad dvc file) and when building a DAG(because user might've created bad dvc file by hand). Take a look at project.graph() method, where we also check for OutputDuplicationError using collected stages. Maybe it is worth leaving these checks where they are and also using them in graph().

**Comment by efiop** [comment]
As @mroutis noticed, that wouldn't actually break anything, because we are using load() to load existing dvc files, and not loads().

I agree about the naming, it is indeed confusing. I was using json's dump/dumps as an inspiration, but it is clearly not too readable. Please feel free to rename as you see fit.

**Pull request merged**